custom_enumerable_methods.rb

```diff
@@ -0,0 +1,119 @@
+module Enumerable
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    my_self = []
+    i = 0
+    while i < my_self.size
+      yield(my_self[i])
+      i += 1
+    end
+    my_self
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    my_self = []
+    (i...my_self.size).my_each do |i|
+      yield(my_self[i], i)
+    end
+    my_self
+  end
```

Comment:
```rb
2.6.5 :028 > my_each_output = ''
2.6.5 :029 > block = proc { |num, idx| my_each_output += "Num: #{num}, idx: #{idx}\n" }
2.6.5 :030 > array.each_with_index(&block)
 => [5, 4, 4, 0, 5, 0, 3, 4, 0, 0, 2, 6, 5, 6, 5, 6, 2, 1, 3, 0, 8, 0, 0, 1, 7, 1, 3, 5, 6, 1, 1, 1, 6, 5, 7, 7, 2, 0, 3, 1, 7, 0, 1, 4, 5, 7, 6, 5, 4, 3, 6, 5, 4, 0, 6, 2, 7, 2, 5, 6, 4, 7, 1, 3, 6, 3, 3, 2, 6, 1, 7, 7, 8, 6, 1, 3, 5, 0, 4, 1, 0, 5, 4, 2, 5, 2, 5, 6, 6, 0, 1, 8, 4, 5, 4, 0, 4, 0, 8, 6] 
2.6.5 :031 > each_output = my_each_output.dup
2.6.5 :032 > my_each_output = ''
2.6.5 :033 > array.my_each_with_index(&block)
Traceback (most recent call last):
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        4: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        3: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        2: from (irb):33
        1: from custom_enumerable_methods.rb:18:in `my_each_with_index'
NameError (undefined local variable or method `i' for #<Array:0x0000000001364b78>)
2.6.5 :034 > 

```

Reply:
```rb
2.6.5 :037 > block = proc { |num| num < 4 }
2.6.5 :038 > array.my_each_with_index(&block)
Traceback (most recent call last):
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        4: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        3: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        2: from (irb):38
        1: from custom_enumerable_methods.rb:18:in `my_each_with_index'
NameError (undefined local variable or method `i' for #<Array:0x0000000001364b78>)
2.6.5 :039 > array.each_with_index(&block)
 => [5, 4, 4, 0, 5, 0, 3, 4, 0, 0, 2, 6, 5, 6, 5, 6, 2, 1, 3, 0, 8, 0, 0, 1, 7, 1, 3, 5, 6, 1, 1, 1, 6, 5, 7, 7, 2, 0, 3, 1, 7, 0, 1, 4, 5, 7, 6, 5, 4, 3, 6, 5, 4, 0, 6, 2, 7, 2, 5, 6, 4, 7, 1, 3, 6, 3, 3, 2, 6, 1, 7, 7, 8, 6, 1, 3, 5, 0, 4, 1, 0, 5, 4, 2, 5, 2, 5, 6, 6, 0, 1, 8, 4, 5, 4, 0, 4, 0, 8, 6] 
2.6.5 :040 >
```

Reply:
# Thank You! :bouquet: 
## Above issue is fixed now! :heavy_check_mark: 

## I helped us to learn more and fix the things.